### PR TITLE
Added new functions to validate fuel types and set output strings

### LINKED ISSUE #7621: Add new function to validate fuel types and set output strings
**This is part 3 of 4: Part 1 = #7532, Part 2 = #7523 , and Part 4 is #7622.**

Issue overview
--------------
The goal of this issue to to remove all if blocks for fuel type in the input processing functions and replace them with a standard function call.

This issue should only standardize the fuel type choice input processing and introduce *no output changes* - that can wait for #?????.

### Background ###
#7523 standardized all fuel type inputs to a single list with no synonyms. Synonyms were removed from all (most?) input processing functions. #7523 introduces no output changes.

Some input processing using the functions `DataGlobalConstants::AssignResourceTypeNum` and `DataGlobalConstants::GetResourceTypeChar` to validate fuel types and set the standard strings for output variables and meter resource types. But many input processing functions still have their own if blocks to test for "NaturalGas" "Electricity" etc. This issue is to use a standard fuel type function everywhere as appropriate. 

### More Details ###
1. Given that fuel types are a subset of resource types, `AssignResourceTypeNum` could be modified to take an argument to only allow fuels (no other resource types) or it may be best to have a new function(s) just for fuel type inputs.

2. In general, objects with a fuel type choice need to validate the input choice and assign strings to use for output variables and the correct resource type to use for meters.

3. `AssignResourceTypeNum` and `GetResourceTypeChar` may be ripe for some refactoring.

4. For this first pass with no output diffs, the mappings are shown below. One might be tempted to say we don't need three strings. That's true for some fuel types, like Diesel, but FuelOilNo1 will ultimately map to "Fuel Oil No 1" for output variables and table headings in the last phase of this. For now some of the old names are kept to produce no diffs.

**Input choice, Meter Resource, Output Variable**
Electricity, Electricity, Electric
NaturalGas, Gas, Gas
Propane, Propane, Propane
Diesel, Diesel, Diesel
Gasoline, Gasoline, Gasoline
FuelOilNo1, FuelOil#1, FuelOil#1
FuelOilNo2, FuelOil#2, FuelOil#2
Coal, Coal, Coal
OtherFuel1, OtherFuel1, OtherFuel1
OtherFuel2, OtherFuel2, OtherFuel2
Water, Water, Water

4. For every type except NaturalGas, the meter resource type is identical to the input fuel type. And ultimately, NaturalGas will be the same, also. So perhaps the input fuel type can simply be passed directly to the `ResourceTypeKey` in `SetupOutputVariable`.

5. Other related functions are "GetStandardMeterResourceType" which is called from during SetupOutputVariable processing. So, this function could stay as-is and take care of the one oddball meter resource "Gas" which ultimately will move to "NaturalGas". This function should be evaluated for refactoring as well. Even though the input choices should be fixed in the idd/schema, this function should probably remain to trap any oddball nonstandard fuel types that might get added. And it can clean up any upper/lower case inconsistencies.

6. There may be some subtleties here that gum up the works for the no-diff case. For example electricity use is typically "Electric Power" and "Electric Energy" in output variables. While most (all?) other fuels are "Gas Rate", "Gas Energy".

Lots of scattered info here - sorry.

### Details
Some additional details for this issue (if relevant):
 - Followup for #5941.

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by dareumnam**
Pull request overview
---------------------
 - Fixes #7621
 - Removed if blocks for fuel type in the input processing functions and replaced them with standard functions calls 
 - The three standard functions were added in UtilityRoutines.cc & UtilityRoutines.hh
 - `bool ValidateFuelType(std::string const &FuelTypeInput,
                          std::string &FuelTypeOutput,
                          bool &FuelTypeErrorsFound)`
Replace if blocks with **fuel type strings** in:

| Object | File | Fuel Type in the IDD |
| --- | --- | --- |
| Chiller:EngineDriven | PlantChillers.cc | \required-field |
| Chiller:CombustionTurbine | PlantChillers.cc | \default was set |
| ChillerHeater:Absorption:DirectFired | ChillerGasAbsorption.cc | \default was set |
| Generator:InternalCombustionEngine | ICEngineElectricGenerator.cc | \default was set |
| Generator:CombustionTurbine | CTElectricGenerator.cc | \default was set  |
| Generator:MicroTurbine | MicroturbineElectricGenerator.cc | \default was set |

 - `bool ValidateFuelTypeWithFuelTypeNum(std::string const &FuelTypeInput,
                                         int &FuelTypeNum,
                                         bool &FuelTypeErrorsFound)`
Replace if blocks with **fuel type numbers** in :

| Object | File | Fuel Type in the IDD |
| --- | --- | --- |
| AirConditioner:VariableRefrigerantFlow | HVACVariableRefrigerantFlow.cc | \default was set |
| Coil:Cooling:DX:MultiSpeed | DXCoils.cc | \required-field |
| Coil:Heating:DX:MultiSpeed | DXCoils.cc | \required-field |

   - `bool ValidateFuelTypeWithAssignResourceTypeNum(std::string const &FuelTypeInput,
                                                   std::string &FuelTypeOutput,
                                                   int &FuelTypeNum,
                                                   bool &FuelTypeErrorsFound)`
Replace if blocks with **fuel type strings** and `DataGlobalConstants::AssignResourceTypeNum()` in:

| Object | File | Fuel Type in the IDD |
| --- | --- | --- |
| Boiler:HotWater | Boilers.cc | \required-field |
| Boiler:Steam | BoilerSteam.cc | \required-field |

 - Add unit tests
   - BoilerHotWater.unit.cc 
   - BoilerSteam.unit.cc
   - ChillerGasAbsorption.unit.cc
   - DXCoils.unit.cc 
   - HVACVariableRefrigerantFlow.unit.cc
   - PlantChillers.unit.cc
   - CTElectricGenerator.unit.cc (Added to \tst\EnergyPlus\unit\CMakeLists.txt)
   - ICEngineElectricGenerator.unit.cc (Added to \tst\EnergyPlus\unit\CMakeLists.txt)
   - MicroturbineElectricGenerator.unit.cc (Added to \tst\EnergyPlus\unit\CMakeLists.txt)

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (mitchute) [inline src/EnergyPlus/DataGlobalConstants.cc]: @Myoldmopar @mjwitte is DataGlobalConstants the right place for this?

**Comment by dareumnam** [comment]
@mitchute This is ready for review now.

File: src/EnergyPlus/BoilerSteam.cc
Comment (rraustad) [inline src/EnergyPlus/BoilerSteam.cc]: Where is DataGlobalConstants::FuelTypeErrorsFound set back to false for the next boiler? If the first boiler has an incorrect fuel type then won't this message print for all boilers?

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (rraustad) [inline src/EnergyPlus/DataGlobalConstants.cc]: Would a switch statement be more efficient for these functions?

File: src/EnergyPlus/DataGlobalConstants.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataGlobalConstants.hh]: I do not understand the need for these global variables.  That new function accepts a `FuelTypeInput` argument.  What is this global variable `FuelTypeInput` for?  I think you need to add two arguments to the function to return back the FuelTypeNum and FuelType values.  No need to add global variables.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (Myoldmopar) [inline src/EnergyPlus/DataGlobalConstants.cc]: Instead of having a global variable for fuel type errors found, I would, for now:
- Make this function a `bool` instead of a `void`
- Return true if the function successfully matches a fuel type, and return false if not -- OR return true if errors were found and false if it succeeds, but just be clear about it.

Then the calling function can handle things appropriately.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (Myoldmopar) [inline src/EnergyPlus/DataGlobalConstants.cc]: As I mentioned in another comment, you need to pass in and out the `FuelType` string and the `FuelTypeNum` integer.  This input string also needs to be a constant reference so that it isn't copied when this function is called.  The new signature should be:

```
bool ValidateFuelTypeWithAssignResourceTypeNum(
    std::string const & FuelTypeInput,
    std::string & FuelType,
    int & FuelTypeNum
);
```

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (Myoldmopar) [reply src/EnergyPlus/DataGlobalConstants.cc]: Can't switch on a string, unfortunately.  In the functions that convert integer to the appropriate string, you can switch on the integers.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (Myoldmopar) [reply src/EnergyPlus/DataGlobalConstants.cc]: I'm not sure.  Maybe UtilityRoutines.cc?

File: tst/EnergyPlus/unit/ICEngineElectricGenerator.unit.cc
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/ICEngineElectricGenerator.unit.cc]: Since we are taking the global variables back out, you won't be able to check this.  And that's OK.  In the GetInput routine, the code that is performing input validation should check the value returned from the fuel validation function, and decide what to do.  If an error occurs, the GetInput routine should wrap up and throw a fatal error.  In that case, you can test successful fuel entries just by calling GetInput on good values.  And then you can test bad fuel entries by using EXPECT_THROW.  You can find examples if you search for that text in the unit test code.

**Review by Myoldmopar — CHANGES_REQUESTED** [review]
The biggest change needed here is to remove the global variables.  Those can be eliminated by just passing extra variables into and out of the new worker functions, and having the function return true/false for success or failure.  Then the calling code needs only to check the return value and decide whether to gracefully throw a fatal or continue on.  Unit tests will need to be updated accordingly.

Other than that, the fuel function code seems fine, I'll give it a final pass once the global variables are taken care of.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (rraustad) [reply src/EnergyPlus/DataGlobalConstants.cc]: Be careful, if there are other errors in GetInput, then this function could over-ride ErrorsFound to false.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (rraustad) [inline src/EnergyPlus/DataGlobalConstants.cc]: Please delete these comment lines with "na" in a future commit. I don't mind the name/date comments (could be 1 line though) but other disagree. So at the very least get rid of the na lines. Keep new functions as slim as possible.

**Comment by rraustad** [comment]
@dareumnam nice job so far!

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (rraustad) [reply src/EnergyPlus/DataGlobalConstants.cc]: I think this warrants another function argument for &ErrorsFound. Where an error is passed back only if true. @jmarrec updated this a while back. See GetFanInletNode as an example.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (rraustad) [inline src/EnergyPlus/DataGlobalConstants.cc]: This is so much better.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (rraustad) [inline src/EnergyPlus/DataGlobalConstants.cc]: So is setting 3 vars efficient? Is FuelTypeNum only used in DXCoils and HVACVariableRefrigeratntFlow? Seems like you need only 1 int here. Is FuelTypeNumforAssignResource and FuelTypeNum the same value? Seems likely.

Comment (dareumnam) [reply src/EnergyPlus/DataGlobalConstants.cc]: FuelTypeNum is used in DXcoils and HVACVariantRefrigerantFlow, and FuelTypeNumforAssignResouce is used in Boilers and Boilersteam with AssignResourceTypeNum() function (https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/DataGlobalConstants.cc#L325-L490), but these two are not the same int value for each fuel type, so once I separated these two int vars. I'll figure out if these can be combined while removing the global vars. Thank you for your insight.

File: src/EnergyPlus/BoilerSteam.cc
Comment (dareumnam) [reply src/EnergyPlus/BoilerSteam.cc]: Thank you for pointing out. I added lines to set the error flag back.

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (dareumnam) [reply src/EnergyPlus/DataGlobalConstants.cc]: I'll move the function to UtilityRoutines.cc!

File: src/EnergyPlus/DataGlobalConstants.cc
Comment (dareumnam) [reply src/EnergyPlus/DataGlobalConstants.cc]: Thanks for your insights. Since C++ can't switch on a string, once I used enum to convert strings to integer and make one standard function instead of the three functions. But I think I might need to return back to if-else blocks while removing global vars.

**Comment by mitchute** [comment]
@Myoldmopar @rraustad It looks like Dareum got this wrapped up and addressed your comments. Do you need any other changes here before this goes in? I'm merging in develop now to let CI take one more pass.

**Review by Myoldmopar — APPROVED** [review]
It's so nice having all this functionality in one location.  The HVACVRF file changes are difficult to mine through because the formatting did so much, but I think that's OK.  As long as CI is happy with it, I trust it didn't do any damage in there.

File: src/EnergyPlus/UtilityRoutines.hh
Comment (mjwitte) [inline src/EnergyPlus/UtilityRoutines.hh:581]: @dareumnam I know this it merged already, but I'm wondering why there are three separate functions here? Seems like these could be collapsed into one function, or at least build off each other so that there's only one place with ifs on the various fuel type strings.

Also, there is still a similar function in [`ExteriorEnergyUse::ValidateFuelType`](https://github.com/NREL/EnergyPlus/blob/8e084f1905da7c4c88baba8b67fcdf3bdd066dab/src/EnergyPlus/ExteriorEnergyUse.cc#L449-L454).

Comment (dareumnam) [reply src/EnergyPlus/UtilityRoutines.hh:581]: @mjwitte Thanks for your comment. This was to remain each object as original as possible, particularly setup output variables and error messages. As you comment, they can be collapsed into a single function if we want to do that, but it needs every object to get fixed. I can do that if it is needed.

Comment (mjwitte) [reply src/EnergyPlus/UtilityRoutines.hh:581]: @dareumnam I would like to finish addressing #7622 before the I/O freeze on Aug 19. The changes for Fuel Oil are waiting for final review in #8134. Then we can build on that and finish up all of the fuel/meter naming work.

Comment (dareumnam) [reply src/EnergyPlus/UtilityRoutines.hh:581]: @mjwitte I see. Then I'll wrap up #7622 right after #8134 gets merged. Thanks for letting me know.

Comment (mjwitte) [reply src/EnergyPlus/UtilityRoutines.hh:581]: @dareumnam Also, note a slight change from the original plan. For output variables we will now use the condensed PascalCase names, not the expanded names. For example "Boiler NaturalGas Energy". This is different from the original plan that was laid out in #7622. The only place expanded names will be used is for table headings (Natural Gas).

**Pull request merged**
